// @ngx-cache/core/esm2015/index.js
import { InjectionToken, Injector, NgModule, Optional, PLATFORM_ID, SkipSelf } from '@angular/core';
import { CacheLoader, CacheStaticLoader } from './src/cache.loader';
import { CacheService } from './src/cache.service';
export * from './src/cache';
export * from './src/cache.loader';
export * from './src/cache.service';
export * from './src/cached.decorator';
export * from './src/storage';
export const STORAGE = new InjectionToken('STORAGE');
export function cacheFactory() {
    return new CacheStaticLoader();
}
export function cacheServiceFactory(loader, platformId, injector) {
    return new CacheService(loader, platformId, injector);
}
export class CacheModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('CacheModule already loaded; import in root module only.');
        }
    }
    static forRoot(configuredProvider = {
        provide: CacheLoader,
        useFactory: cacheFactory
    }) {
        return {
            ngModule: CacheModule,
            providers: [
                configuredProvider,
                {
                    provide: CacheService,
                    useFactory: cacheServiceFactory,
                    deps: [CacheLoader, PLATFORM_ID, Injector]
                }
            ]
        };
    }
}
CacheModule.decorators = [
    { type: NgModule }
];
CacheModule.ctorParameters = () => [
    { type: CacheModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
